Sources/Sentry/SentrySessionReplay.m

```diff
@@ -0,0 +1,264 @@
+#import "SentrySwift.h"
+#import "SentryViewPhotographer.h"
+
+#if SENTRY_HAS_UIKIT
+
+static NSString *SENTRY_REPLAY_FOLDER = @"replay";
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation SentrySessionReplay {
+    UIView *_rootView;
+    BOOL _processingScreenshot;
+    CADisplayLink *_displayLink;
+    NSDate *_lastScreenShot;
+    NSDate *_videoSegmentStart;
+    NSURL *_urlToCache;
+    NSDate *_sessionStart;
+    SentryReplayOptions *_replayOptions;
+    SentryOnDemandReplay *_replayMaker;
+    SentryId *sessionReplayId;
+    NSMutableArray<UIImage *> *imageCollection;
+    int _currentSegmentId;
+    BOOL _isFullSession;
+}
+
+- (instancetype)initWithSettings:(SentryReplayOptions *)replayOptions
+{
+    if (self = [super init]) {
+        _replayOptions = replayOptions;
+    }
+    return self;
+}
+
+- (SentryCurrentDateProvider *)dateProvider
+{
+    return SentryDependencyContainer.sharedInstance.dateProvider;
+}
+
+- (void)start:(UIView *)rootView fullSession:(BOOL)full
+{
+    if (rootView == nil) {
+        SENTRY_LOG_DEBUG(@"rootView cannot be nil. Session replay will not be recorded.");
+        return;
+    }
+
+    @synchronized(self) {
+        if (_displayLink == nil) {
+            _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(newFrame:)];
+            [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
+        } else {
+            // Session display is already running.
+            return;
+        }
+
+        _rootView = rootView;
+        _lastScreenShot = [[NSDate alloc] init];
+        _videoSegmentStart = nil;
+        _sessionStart = _lastScreenShot;
+        _currentSegmentId = 0;
+
+        NSURL *docs = [NSURL
+            fileURLWithPath:[SentryDependencyContainer.sharedInstance.fileManager sentryPath]];
+        docs = [docs URLByAppendingPathComponent:SENTRY_REPLAY_FOLDER];
+
+        NSString *currentSession = [NSUUID UUID].UUIDString;
+        _urlToCache = [docs URLByAppendingPathComponent:currentSession];
+
+        if (![NSFileManager.defaultManager fileExistsAtPath:_urlToCache.path]) {
+            [NSFileManager.defaultManager createDirectoryAtURL:_urlToCache
+                                   withIntermediateDirectories:YES
+                                                    attributes:nil
+                                                         error:nil];
+        }
+
+        _replayMaker = [[SentryOnDemandReplay alloc] initWithOutputPath:_urlToCache.path];
+        _replayMaker.bitRate = _replayOptions.replayBitRate;
+        _replayMaker.cacheMaxSize = (NSInteger)(full ? _replayOptions.sessionSegmentDuration
+                                                     : _replayOptions.errorReplayDuration);
+        imageCollection = [NSMutableArray array];
+
+        NSLog(@"Recording session to %@", _urlToCache);
+
+        _isFullSession = full;
+        if (full) {
+            sessionReplayId = [[SentryId alloc] init];
```

Comment:
we need a new session id everytime we end/start etc, should this be in some function like "resetReplayId" and invoke that from here?